src/Swift.Bindings/src/Model/TypeSpec.cs

```diff
@@ -0,0 +1,380 @@
+	public bool IsInOut { get; set; }
+	public bool IsAny { get; set; }
+	public virtual bool IsEmptyTuple { get { return false; } }
+	protected abstract string LLToString(bool useFullName);
+	protected virtual string LLFinalStringParts() { return ""; }
+	protected abstract bool LLEquals(TypeSpec? other, bool partialNameMatch);
+	public string? TypeLabel { get; set; }
+	public bool IsArray
+	{
+		get
+		{
+			return this is NamedTypeSpec ns && ns.Name == "Swift.Array";
+		}
+	}
+
+	public override bool Equals(object? obj)
+	{
+		if (obj is TypeSpec spec)
+		{
+			if (Kind != spec.Kind)
+				return false;
+			if (!ListEqual(GenericParameters, spec.GenericParameters, false))
+				return false;
+			if (IsInOut != spec.IsInOut)
+				return false;
+			// Don't compare IsAny - it's really not important (yet)
+			return LLEquals(spec, false);
+		}
+		return false;
+	}
+
+	public bool EqualsPartialMatch(TypeSpec spec)
+	{
+		if (spec == null)
+			return false;
+		if (Kind != spec.Kind)
+			return false;
+		if (!ListEqual(GenericParameters, spec.GenericParameters, true))
+			return false;
+		if (IsInOut != spec.IsInOut)
+			return false;
+		// Don't compare IsAny - it's really not important (yet)
+		return LLEquals(spec, true);
+	}
+
+	public virtual bool EqualsReferenceInvaraint(TypeSpec type)
+	{
+		var a = ProjectAsNonReference(this);
+		var b = ProjectAsNonReference(type);
+
+		if (b.Kind != a.Kind)
+			return false;
+		if (b.GetType() != a.GetType())
+			return false;
+		// shouldn't do Name equality except in functions
+		return a.LLEquals(b, false);
+	}
+
+	public TypeSpec NonReferenceCloneOf()
+	{
+		if (!IsInOut)
+			return this;
+		var ty = MemberwiseClone() as TypeSpec;
+		ty!.IsInOut = false;
+		return ty!;
+	}
+
+	static TypeSpec ProjectAsNonReference(TypeSpec a)
+	{
+		if (a.IsInOut)
+		{
+			return a.NonReferenceCloneOf();
+		}
+		if (a is NamedTypeSpec namedType && namedType.GenericParameters.Count == 1)
+		{
+			if (namedType.Name == "Swift.UnsafePointer" || namedType.Name == "Swift.UnsafeMutablePointer")
+				return namedType.GenericParameters[0];
+		}
+		return a;
+	}
+
+	public override int GetHashCode()
+	{
+		return ToString().GetHashCode();
+	}
+
+	protected static bool ListEqual(List<TypeSpec> one, List<TypeSpec> two, bool partialNameMatch)
+	{
+		if (one.Count != two.Count)
+			return false;
+		for (int i = 0; i < one.Count; i++)
+		{
+			if (partialNameMatch)
+			{
+				if (!one[i].EqualsPartialMatch(two[i]))
+					return false;
+			}
+			else
+			{
+				if (!one[i].Equals(two[i]))
+					return false;
+			}
+		}
+		return true;
+	}
+
+	public static bool IsNullOrEmptyTuple(TypeSpec? spec)
+	{
+		return spec is null || spec.IsEmptyTuple;
+	}
+
+	public static bool BothNullOrEqual(TypeSpec? one, TypeSpec? two)
+	{
+		if (one is null && two is null)
+			return true;
+		if (one is null || two is null)
+			return false;
+		return one.Equals(two);
+	}
+
+	public bool ContainsBoundGenericClosure()
```

Comment:
Perhaps we can move this to ClosureTypeSpec?

Reply:
No - this drills down into all the components of all type specs. A `NamedTypeSpec` can contain a bound generic closure in its generic parameters, for example. I don't see an immediate need for this one, so I think I'll take it out entirely.